### DISCUSSION TITLE
Resetting Testnet: Bitcoin Explorers, Indexers & APIs

### DISCUSSION THREAD

**Discussion by CharlieC3** [stacks-network/stacks-core#5578] [General]
Today we discussed the pros/cons of running a Stacks Testnet against a private Bitcoin Regtest network. One of the challenges we're looking to understand is the difficulties and feature-gaps of running a Bitcoin explorer, indexer, and API which other entities and groups in the Stacks ecosystem may rely upon, as many of these services are available on the internet for Bitcoin mainnet (e.g. https://mempool.space/).

Some folks at Hiro, such as myself, @zone117x and @saralab, researched a bunch of different OSS in this field in early 2023. Our research led us to understand that the [mempool](https://github.com/mempool/mempool) frontend & backend projects, along with the [mempool electrs indexer/API](https://github.com/mempool/electrs) had the best feature-set and alleged performance and reliability at the time.
This is what we're using to power https://mempool.bitcoin.regtest.hiro.so/

To better understand the challenges in running these services and their feature-gaps, I thought it would be useful to provide the research we conducted in early 2023 to facilitate a discussion around this matter in hopes that it helps us make a more informed decision on what the future of the Stacks Testnet network may look like.

It should be noted that:
* Many of the different projects in the Bitcoin explorer/api/indexer space are often forks of each other with usually minor differences between them. So if there are incompatibility issues found in one of our selections, replacing the component with a different project may not resolve any compatibility issues as they may be based on the same code, or may introduce other challenges that the original project accounts for, such as performance or reliability concerns.
* As this research was conducted in early 2023, take it with a grain of salt and when in doubt verify it again as the status for some of these projects may have changed.
* The [mempool electrs indexer/API that we chose](https://github.com/mempool/electrs) is not mentioned in the list below, as it was created between the time we finished this research and implemented it. But since it was based on [romanz/electrs](https://github.com/romanz/electrs) and [Blockstream/electrs](https://github.com/Blockstream/electrs) which our research favored, it checked all the boxes and made the most sense to use with the mempool frontend & backend projects.

# Bitcoin Indexer Comparison (early 2023)

| Project | Protocol | Prod-ready | Deployment Scope | Scalability | Metrics support | Community Support | Verdict |
| --- | --- | --- | --- | --- | --- | --- | --- |
| [`romanz/electrs`](https://github.com/romanz/electrs) | ⚠️ [`Electrum`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol.html) | ❌<br>Note that this implementation of Electrum server is optimized for personal/small-scale (family/friends) usage. It's a bad idea to run it publicly as it'd expose you to DoS and maybe also other attacks. | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires an empty disk to store database locally | ❌<br>Does not scale well; [scaling up would require a separate database.](https://github.com/romanz/electrs/blob/master/doc/config_example.toml#L21)<br><br>This could be done with snapshots potentially, but is time-consuming and increases overhead.<br>Load balancing may return differing data with session persistence. | [✅](https://github.com/romanz/electrs/blob/master/doc/monitoring.md#monitoring) (Prometheus) | ✅<br>> 500 stars<br>Latest official release: 1/5/23 | ❌ 1<br>Intended for small-scale operations |
| [`blockstream/electrs`](https://github.com/blockstream/electrs) | ✅ [`Electrum`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol.html) + [`Esplora`](https://github.com/blockstream/esplora/blob/master/API.md) | ✅<br>Used as the backend for the Esplora block explorer powering blockstream.info. | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires an empty disk to store database locally | ❌<br>Does not scale well; [scaling up would require a separate database.](https://github.com/romanz/electrs/blob/master/doc/config_example.toml#L21)<br><br>This could be done with snapshots potentially, but is time-consuming and increases overhead.<br><br>Load balancing may return differing data with session persistence. | [✅](https://github.com/Blockstream/electrs/blob/new-index/doc/usage.md#monitoring) (Prometheus) | ⚠️<br>~200 stars<br>Latest official release: 10/14/18<br>Latest commit to default branch: 6/22/22<br>The GH issues page has been hidden; no clear way to submit new issues to track current ones | ⚠️ 3<br>Flexible due to combination of protocols<br>May need to change the storage engine to improve scalability. |
| [`coinbase/rosetta-bitcoin`](https://github.com/coinbase/rosetta-bitcoin) | ✅ [`Rosetta`](https://www.rosetta-api.org/docs/Reference.html) | ❌<br>ROSETTA-BITCOIN IS CONSIDERED ALPHA SOFTWARE. | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires an empty disk to store database locally | ❌<br>Does not scale well; scaling up would require a separate database.<br><br>This could be done with snapshots potentially, but is time-consuming and increases overhead.<br><br>Load balancing may return differing data with session persistence. | ❌ | ⚠️<br>~90 stars<br>Latest official release: 2/8/22<br>Latest commit to default branch: 11/1/22 | ❌ 1<br>Coinbase is not running this and has not maintained it for a while |
| [`dpc/rust-bitcoin-indexer`](https://github.com/dpc/rust-bitcoin-indexer) | ✅ Custom ([Postgres](https://github.com/dpc/rust-bitcoin-indexer/blob/master/src/db/schema.rs)) | ⚠️<br>An experiment in creating a perfect Bitcoin Indexer ... This started as an educational experiment, but is quite advanced already. ... rust-bitcoin-indexer (at least right now) is just a neat experiment that gave good results and seems to be working quite well. | ✅<br>Hiro would need to build and maintain a Docker image<br><br>Requires a Postgres database deployed | ⚠️<br>Uses a Postgres db, so the app can theoretically scale horizontally. Since this project is not prod-ready, we’d still need to verify how well the app performs under heavy load. | ❌ | ❌<br>~90 stars<br>No official releases/tags<br>Latest commit to default branch: 5/17/21 | ❌ 1<br>Recognized as an “experiment”, low dev activity |
| [`cculianu/Fulcrum`](https://github.com/cculianu/Fulcrum) | ⚠️ [`Electrum`](https://electrum-cash-protocol.readthedocs.io/en/latest/) (Fulcrum fork) | ❔<br>unclear | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires an empty disk to store database locally | ❌<br>Does not scale well; [scaling up would require a separate database.](https://github.com/cculianu/Fulcrum/blob/master/doc/fulcrum-example-config.conf#L48)<br><br>This could be done with snapshots potentially, but is time-consuming and increases overhead.<br><br>Load balancing may return differing data with session persistence. | ❌ | ✅<br>~160 stars<br>Latest official release: 12/20/22<br>Active development, issues being resolved regularly | ❌ 2<br>Poor scaling capabilities, Electrum protocol has limitations |
| [`spesmilo/electrumx`](https://github.com/spesmilo/electrumx) | ⚠️ [`Electrum`](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol.html) | ❔<br>unclear | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires an empty disk to store database locally | ❌<br>Does not scale well; [scaling up would require a separate database.](https://github.com/spesmilo/electrumx/blob/master/docs/HOWTO.rst#database-engine)<br><br>This could be done with snapshots potentially, but is time-consuming and increases overhead.<br><br>Load balancing may return differing data with session persistence. | ❌ | ⚠️<br>~300 stars<br>Latest official release: 12/10/20<br>Latest commit to default branch: 11/10/22<br><br>Despite no official release/tag in a while, there’s still plenty of active development on this project<br> | ❌ 2<br>Poor scaling capabilities, Electrum protocol has limitations |
| [`block-core/blockcore-indexer`](https://github.com/block-core/blockcore-indexer) | ✅ [`Blockcore`](https://btc.indexer.coinvault.io/docs/index.html) | ✅<br>Used as the backend for the [Blockcore.net explorer](https://explorer.blockcore.net/btc/api) | ✅<br>[Official Docker image exists](https://hub.docker.com/r/blockcore/indexer)<br><br>Requires a MongoDB database deployed | ⚠️<br>Uses MongoDB, so the app can theoretically scale horizontally better than an app using a local directory can.<br><br>Since this project is used by Blockcore.net, chances are it has some capability to scale.<br><br>It seems [blockcore.net](http://blockcore.net) is a personal project or doesn’t receive high traffic. Scaling exploration would need to be carried out | ❌ | ⚠️<br>17 stars<br>Latest official release: 12/16/22<br>Latest commit to default branch: 12/21/22<br><br>Seems like a small team actively working on this project; issues getting resolved at a slow pace. If Hiro needs a new feature added to bug fixed, we might need to contribute.<br><br>Written in C#; Hiro doesn’t have many C# devs besides Matt 🙂 | ⚠️ 2.5<br>Low community usage<br>Hiro isn’t historically a C# shop<br>Low confidence in its usage for a low traffic site |
| [`bitauth/chaingraph`](https://github.com/bitauth/chaingraph) | ✅ Custom ([GraphQL](https://chaingraph.cash/) + [Postgres](https://github.com/bitauth/chaingraph/blob/master/docs/schema.md)) | ❌<br>[Limited BTC support (was designed for BCH)](https://github.com/bitauth/chaingraph/issues/36) | ✅<br>[Official Docker images exist](https://hub.docker.com/search?q=%20chaingraph)<br><br>Requires a Postgres database deployed | ⚠️<br>Uses a Postgres db, so the app can theoretically scale horizontally. Since this project is not prod-ready, we’d still need to verify how well the app performs under heavy load.<br><br>Also has a [K8s Helm chart](https://github.com/bitauth/chaingraph/tree/master/charts/chaingraph), making is simple to deploy internally. | ⚠️<br>[Really limited and custom metrics](https://github.com/bitauth/chaingraph/blob/e9c37a6f79fe5d0dc2e0c88cc3a654a3b0d9062e/src/components/throughput-statistics.ts). Might be able to find a way to stream this into Prometheus | ⚠️<br>24 stars<br>Latest official release: 5/14/22<br>Latest commit to default branch: 1/6/23<br><br>Despite some active development, issues are *not* getting resolved at all | ❌ 1<br>Limited/questionable BTC functionality<br>BTC support is unofficial |
| [`bitpay/bitcore`](https://github.com/bitpay/bitcore/tree/master/packages/bitcore-node) | ✅ Custom ([REST](https://github.com/bitpay/bitcore/blob/master/packages/bitcore-node/docs/api-documentation.md)) | ✅<br>Used as the backend for the [Bitpay.com Explorer](https://bitpay.com/insight/BTC/mainnet/blocks) | ⚠️<br>Hiro would need to build and maintain a Docker image<br><br>Requires a MongoDB database deployed<br><br>[Some endpoints may need to be blocked.](https://github.com/bitpay/bitcore/blob/master/packages/bitcore-node/docs/api-documentation.md#add-wallet) | ✅<br>Uses MongoDB, and [supports specification of the db connection details in the config](https://github.com/bitpay/bitcore/blob/master/packages/bitcore-node/src/types/Config.ts#L54), so the app can theoretically scale horizontally.<br><br>Since this project is used by Bitpay.com, chances are it has some capability to scale. | ⚠️<br>[Really limited and custom metrics](https://github.com/bitpay/bitcore/blob/master/packages/bitcore-node/docs/api-documentation.md#stats). Might be able to find a way to stream this into Prometheus | ✅<br>4.5k stars<br>Latest official release:<br>[10/16/22 (npm)](https://www.npmjs.com/package/bitcore-node)<br>[1/2/23 (gh tag)](https://github.com/bitpay/bitcore/tags)<br>Active development, issues being resolved regularly<br> | ⚠️ 3.5<br>Good dev activity and community support<br>Used in high(ish) traffic Explorer<br>Scales with remote database |
| [`mempool/mempool`](https://github.com/mempool/mempool) | ✅ Custom ([REST](https://mempool.space/docs/api/rest))<br><br>[Integrates with a backend Electrum or Esplora server to increase functionality.](https://github.com/mempool/mempool/tree/master/backend#3-configure-electrum-server) | ✅<br>Used as the backend for the [mempool.space](https://mempool.space) | ✅<br>[Official Docker images exist](https://hub.docker.com/r/mempool/backend)<br><br>[May require deployment of an Electrum server for increased functionality](https://github.com/mempool/mempool/tree/master/backend#3-configure-electrum-server). Seems to be compatible with romanz/electrs or cculianu/Fulcrum and Blockstream/electrs.<br><br>[Lots of configuration suggestions for deploying to production.](https://github.com/mempool/mempool/tree/master/production) | ✅<br>Uses MariaDB, and [supports specification of the db connection details in the config](https://github.com/mempool/mempool/blob/master/backend/mempool-config.sample.json#L52), so the app can theoretically scale horizontally.<br><br>Since this project is used by mempool.space, chances are it has some capability to scale. | ❌ | ✅<br>1.2k stars<br>Latest official release/tag:<br>7/25/22<br><br>Despite no new release in 6 months, there’s active development almost every day against `master`, issues being resolved regularly<br> | ✅<br>Charlie: 4<br>Matt: 4<br>Sarala : 4 |<br>
### Protocol usability

The `Electrum` protocol has a limitation that makes fetching txs for a given block pretty expensive (see [`blockchain.transaction.id_from_pos`](https://electrumx.readthedocs.io/en/latest/protocol-methods.html#blockchain-transaction-id-from-pos)).

This may be a considered a downside for the following projects: `romanz/electrs`, `cculianu/Fulcrum`, `spesmilo/electrumx`.

---

**Comment by jcnelson** [comment]
Taking notes from the call:

* A strong point against testnet4 is that it still encounters block storms, which makes it unsuitable for hosting Nakamoto (since testnet4 could produce an empty prepare phase)

* Signet may not have block storms, but we need to check.  Is it a viable alternative?
   * It definitely has flash blocks though
   * It would require potentially-substantial code changes in the Stacks node to support
   * It still forces users to run a signet bitcoind node and somehow obtain coins for it

* A strong point against hosting a testnet inhouse is that we have to solve the problem of tBTC running out without resetting the testnet. 
   * Running a fork of bitcoind where this isn't a problem requires all partners to run our forked code
   * We have to maintain the fork
   
* Running a testnet Stacks miner requires additional services to operate on an inhouse testnet (but these all seem solvable):
   * It would require us to add a way for them to register their mining key(s) in our hosted bitcoind node, if they don't run one themselves
      * Probably not something we need to do right now, since people can already peer their own bitcoind nodes
   * We would need to solely control block mining
      * Already do this
   * It would require us to host and maintain a tBTC faucet (this is already done!), and defend it from griefers 
      * The faucet could be much smaller, and we could make our miners mine less
      * Our stackers should treat the faucet address as their PoX rewards address
      * We should publish the testnet faucet address as a sink, so people can return their testnet BTC
   * We would need to support users peering their regtest bitcoind nodes 
   * Slower block times to save BTC?

* Set a schedule for nuking public testnet?

* Write docs for setting up a local "testnet2" so we can iterate and test all of the above

Consensus at the end of the call seems to be to keep our current setup as-is, but iterate on it to make it more robust.  No concrete actions are to be taken until January, given the upcoming holiday.

**Reply by CharlieC3** [reply]
> No concrete actions are to be taken until January, given the upcoming holiday

FYI @aldur

**Reply by aldur** [reply]
Thanks both! 

Speaking from the experience of the sBTC launch: it has been very challenging for downstream users (e.g., wallets) to integrate sBTC because of the lack of a stable testnet environment. 

We _seem_ to have one working right now (running `3.1.0.0.1-rc2`). If that's the case, I'd be OK leaving things as they are through the holidays. 

However, if the deployment we have turns out to be brittle and continues creating challenges for builders, then I'd advocate for an immediate reset of the testnet. This is of course just a band-aid, giving builders a way to integrate sBTC _now_ when momentum requires it. We would still need to iterate and make more robust later.

**Reply by aldur** [reply]
> * Set a schedule for nuking public testnet?

Optimizing for simplicity, mid-term this seems to be a good option.

**Reply by wileyj** [reply]
One other thing that i'd like for us to think about - a contingency plan for the single bitcoin regtest rpc miner. 
not a blocker to reset, and i think we're all in agreement that the issues raised in this discussion can be addressed afterwards, but i would also like to add a contingency for the miner specifically - either as a hot/cold standby miner (probably the better option), or we alternate who mines blocks (this would be difficult though, since timing could prove challenging). 

i'll dig into some ideas here, the last time i looked there weren't any *good* options that would allow decentralized miners.
